[Style] Remove unnecessary inheritance from OutlineValue type

weinig opened the pull request:
#### 5c53402ab6e8237a8a170a8b12aa15c4ba66fd34
<pre>
[Style] Remove unnecessary inheritance from OutlineValue type
<a href="https://bugs.webkit.org/show_bug.cgi?id=295488">https://bugs.webkit.org/show_bug.cgi?id=295488</a>

Reviewed by Darin Adler.

The `OutlineValue` type previously inherited from the `BorderValue` type, forcing
outlines to use `BorderStyle` to represent their style, despite `BorderStyle` having
invalid styles, like `hidden`, and missing `auto`, which was bolted on with a bool.

This removes the inheritance and adds a separate `OutlineStyle` enum to fully represent
the `outline-style` property&apos;s value space.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add OutlineValue.cpp

* Source/WebCore/css/CSSPrimitiveValueMappings.h:
    - Remove special casing of auto in BorderStyle mapping and adds OutlineStyle mapping.

* Source/WebCore/rendering/BorderEdge.cpp:
* Source/WebCore/rendering/BorderEdge.h:
* Source/WebCore/rendering/BorderPainter.cpp:
    - To account for outline style no longer using BorderStyle, we now have to
      conditionally convert to BorderStyle for painting.

* Source/WebCore/rendering/RenderBlock.cpp:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderInline.cpp:
* Source/WebCore/rendering/RenderObject.cpp:
* Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
    - Replace calls to `RenderStyle::hasAutoOutlineStyle()` with `RenderStyle::outlineStyle() == OutlineStyle::Auto`.

* Source/WebCore/rendering/style/BorderData.cpp:
* Source/WebCore/rendering/style/BorderData.h:
    - Move `BorderValue` and `OutlineValue` logging to their respective files.
    - Use simpler, but equivalent, defaulting operator==.

* Source/WebCore/rendering/style/BorderValue.cpp:
* Source/WebCore/rendering/style/BorderValue.h:
    - Move trivial constructor to header.
    - Move logging function here from BorderData.h
    - Use simpler, but equivalent, defaulting operator==.
    - Remove `setColor` function, only caller was `RenderStyle` which is already friended.
    - Remove `isAuto` bool, only needed for OutlineValue.
    - Added PREFERRED_TYPE annotation.

* Source/WebCore/rendering/style/OutlineValue.cpp: Added.
* Source/WebCore/rendering/style/OutlineValue.h:
    - Remove inheritance, copying relevant members here.
    - Add `toBorderStyle` conditional conversion function.

* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
    - Switch to using OutlineStyle.
    - Remove special casing for border color setters.

* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
* Source/WebCore/rendering/style/RenderStyleConstants.h:
    - Add OutlineStyle enum.

* Source/WebCore/style/StyleBuilderCustom.h:
    - Remove now unnecessary custom builder code for `outline-style`.

Canonical link: <a href="https://commits.webkit.org/297076@main">https://commits.webkit.org/297076@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69828590269e9e69fc70c6167f171c9e394bd7e9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110479 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30138 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20571 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116504 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/60738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112442 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30817 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38726 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/84020 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/60738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113427 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24595 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99481 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64461 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23958 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17622 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60300 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93979 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17678 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119295 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37519 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/27849 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/92986 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37893 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95750 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92809 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23644 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37807 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15546 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/33476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37414 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42885 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37076 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40416 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38785 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/style/OutlineValue.h:
```diff
@@ -24,20 +25,67 @@
 
 private:
+
+    Style::Color m_color { Style::Color::currentColor() };
```
```suggestion
    Style::Color m_color { Style::Color::currentColor() };
```

darinadler commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -816,7 +819,6 @@ class RenderStyle final : public CanMakeCheckedPtr<RenderStyle> {
     OptionSet<HangingPunctuation> hangingPunctuation() const;
 
-    float outlineOffset() const;
 
```
Leaving an extra blank line here?

weinig replied:
Not intentional.

darinadler commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -325,7 +325,7 @@ inline bool RenderStyle::hasInset() const { return !Style::isZero(insetBox()); }
 inline bool RenderStyle::hasOpacity() const { return m_nonInheritedData->miscData->hasOpacity(); }
 inline bool RenderStyle::hasOutOfFlowPosition() const { return position() == PositionType::Absolute || position() == PositionType::Fixed; }
-inline bool RenderStyle::hasOutline() const { return outlineStyle() > BorderStyle::Hidden && outlineWidth() > 0; }
+inline bool RenderStyle::hasOutline() const { return outlineStyle() != OutlineStyle::None && outlineWidth() > 0; }
```
So much better

darinadler commented on Source/WebCore/rendering/style/RenderStyleSetters.h:
```diff
@@ -32,7 +33,6 @@ namespace WebCore {
 
 #define SET_STYLE_PROPERTY_BASE(read, value, write) do { if (!compareEqual(read, value)) write; } while (0)
 #define SET_STYLE_PROPERTY(read, write, value) SET_STYLE_PROPERTY_BASE(read, value, write = value)
-#define SET_STYLE_PROPERTY_WITH_FUNCTIONS(read, write, value) SET_STYLE_PROPERTY_BASE(read(), value, write(value))
```
So much better

darinadler commented on Source/WebCore/rendering/style/RenderStyleSetters.h:
```diff
@@ -44,9 +44,6 @@ namespace WebCore {
 #define SET_NESTED_PAIR(group, parent, variable1, value1, variable2, value2) SET_STYLE_PROPERTY_PAIR(group->parent, group.access().parent.access(), variable1, value1, variable2, value2)
 #define SET_DOUBLY_NESTED_PAIR(group, grandparent, parent, variable1, value1, variable2, value2) SET_STYLE_PROPERTY_PAIR(group->grandparent->parent, group.access().grandparent.access().parent.access(), variable1, value1, variable2, value2)
 
-#define SET_BORDER_COLOR(group, variable, value) SET_STYLE_PROPERTY_WITH_FUNCTIONS(group->variable.color, group.access().variable.setColor, value)
-#define SET_NESTED_BORDER_COLOR(group, parentVariable, variable, value) SET_STYLE_PROPERTY_WITH_FUNCTIONS(group->parentVariable->variable.color, group.access().parentVariable.access().variable.setColor, value)
```
So much better

darinadler commented on Source/WebCore/rendering/style/RenderStyleSetters.h:
```diff
@@ -690,17 +686,13 @@ inline void RenderStyle::setIsolation(Isolation isolation) { SET_NESTED(m_nonInh
 #undef SET_DOUBLY_NESTED_PAIR
 #undef SET_NESTED
-#undef SET_NESTED_BORDER_COLOR
-#undef SET_NESTED_BORDER_VALUE_COLOR
```
Funny that this was still here

weinig replied:
Yeah, it was bizarre.

weinig committed:
[Style] Remove unnecessary inheritance from OutlineValue type
https://bugs.webkit.org/show_bug.cgi?id=295488

Reviewed by Darin Adler.

The `OutlineValue` type previously inherited from the `BorderValue` type, forcing
outlines to use `BorderStyle` to represent their style, despite `BorderStyle` having
invalid styles, like `hidden`, and missing `auto`, which was bolted on with a bool.

This removes the inheritance and adds a separate `OutlineStyle` enum to fully represent
the `outline-style` property's value space.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add OutlineValue.cpp

* Source/WebCore/css/CSSPrimitiveValueMappings.h:
    - Remove special casing of auto in BorderStyle mapping and adds OutlineStyle mapping.

* Source/WebCore/rendering/BorderEdge.cpp:
* Source/WebCore/rendering/BorderEdge.h:
* Source/WebCore/rendering/BorderPainter.cpp:
    - To account for outline style no longer using BorderStyle, we now have to
      conditionally convert to BorderStyle for painting.

* Source/WebCore/rendering/RenderBlock.cpp:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderInline.cpp:
* Source/WebCore/rendering/RenderObject.cpp:
* Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
    - Replace calls to `RenderStyle::hasAutoOutlineStyle()` with `RenderStyle::outlineStyle() == OutlineStyle::Auto`.

* Source/WebCore/rendering/style/BorderData.cpp:
* Source/WebCore/rendering/style/BorderData.h:
    - Move `BorderValue` and `OutlineValue` logging to their respective files.
    - Use simpler, but equivalent, defaulting operator==.

* Source/WebCore/rendering/style/BorderValue.cpp:
* Source/WebCore/rendering/style/BorderValue.h:
    - Move trivial constructor to header.
    - Move logging function here from BorderData.h
    - Use simpler, but equivalent, defaulting operator==.
    - Remove `setColor` function, only caller was `RenderStyle` which is already friended.
    - Remove `isAuto` bool, only needed for OutlineValue.
    - Added PREFERRED_TYPE annotation.

* Source/WebCore/rendering/style/OutlineValue.cpp: Added.
* Source/WebCore/rendering/style/OutlineValue.h:
    - Remove inheritance, copying relevant members here.
    - Add `toBorderStyle` conditional conversion function.

* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
    - Switch to using OutlineStyle.
    - Remove special casing for border color setters.

* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
* Source/WebCore/rendering/style/RenderStyleConstants.h:
    - Add OutlineStyle enum.

* Source/WebCore/style/StyleBuilderCustom.h:
    - Remove now unnecessary custom builder code for `outline-style`.

Canonical link: https://commits.webkit.org/297076@main
